plugins/README.adoc

```diff
@@ -7,3 +7,256 @@ NOTE: Work in Progress...
+MyPy is a static type checker/
+
+Examples::
+[source,python]
+----
+def greeting(name: str) -> str:
+    return 'Hello ' + name
+----
+
+====
+
+== The use of unittest is discouraged, use pytest instead.
+[%collapsible]
+====
+Explanations:: Use pytest for writing unit tests for plugins
+
+Rationale:: Pytest is the testing framework used by Ansible Engineering and will provide the best experience for plugin developers
+
+Examples::
+[source,python]
+----
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import pytest
+
+from ansible.modules.copy import AnsibleModuleError, split_pre_existing_dir
+from ansible.module_utils.basic import AnsibleModule
+
+ONE_DIR_DATA = (('dir1',
+                 ('.', ['dir1']),
+                 ('dir1', []),
+                 ),
+                ('dir1/',
+                 ('.', ['dir1']),
+                 ('dir1', []),
+                 ),
+                ) 
+
+@pytest.mark.parametrize('directory, expected', ((d[0], d[2]) for d in ONE_DIR_DATA))
+def test_split_pre_existing_dir_one_level_exists(directory, expected, mocker):
+    mocker.patch('os.path.exists', side_effect=[True, False, False])
+    split_pre_existing_dir(directory) == expected
+
+----
+
+====
+
+== A complex plugin argspec should be derived from its documentation runtime if the plugin’s source is manually maintained.
+[%collapsible]
+====
+Explanations::
+A complex plugin argspec should be derived from its documentation runtime if the plugin’s source is manually maintained.  ( >30 keys or depth>2) (action plugins and control node execution modules)
+
+Rationale::
+Hand-crafting large and complex argument specifications can be error-prone and difficult to maintain.  Programmatically generating the spec from DOCUMENTATION (already a structured data object) ensures consistency and accuracy. 
+
+====
+
+== Formatting of manually maintained plugin argspecs
+[%collapsible]
+====
+Explanations::
+Ensure a consistent approach to the way commplex argument_specs are formatted within a collection.
+
+Rationale::
+When hand-writing a complex argspec, the author may choose to build up to data structure from multiple dictionaries or vars.
+Other authors may choose to implement a complex, nested argspec as a single dictionary.
+Within a single collection, select one style and use it consistently.
+
+Examples::
+Single dictionary:
+https://github.com/ansible-collections/cisco.nxos/blob/3.0.0/plugins/module_utils/network/nxos/argspec/bgp_global/bgp_global.py
+
+Multiple dictionaries:
+https://github.com/ansible-collections/community.aws/blob/stable-3/plugins/modules/ec2_scaling_policy.py#L355-L370
+https://github.com/ansible-collections/amazon.cloud/blob/0.1.0/plugins/modules/backup_report_plan.py#L182-L234
+====
+
+
+== A plugin’s entry file should not contain more than 500 lines
```

Comment:
I would keep the concrete number out of the title and explanation, and only keep it in the rationale. You could write that it needs to have a "handy size", and then you can avoid "should" :smirk: